src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14392,217 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
+			goto cleanup;
+
+		g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
+	} else {
+		entrypoints = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+		if (!entrypoints) {
+			aot_printerrf (acfg, "Failed to allocate new entrypoints HashTable.\n");
+			success = FALSE;
+			goto cleanup;
+		}
+		g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
+		g_hash_table_insert (acfg->direct_pinvokes, *module_name_ptr, entrypoints);
+	}
+
+	entrypoints = NULL;
+cleanup:
+	g_free (*module_name_ptr);
+	g_free (*entrypoint_name_ptr);
+	*module_name_ptr, module_name_ptr = NULL;
+	*entrypoint_name_ptr, entrypoint_name_ptr = NULL;
+	return success;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// process_specified_direct_pinvokes processes the direct pinvokes and direct pinvoke lists
+// the user specifies in the direct-pinvokes and direct-pinvoke-lists options and adds the
+// entire module or module and set of entrypoints to the MonoAotCompile instance's
+// direct_pinvoke HashTable.
+//
+// Format of direct_pinvoke HashTable:
+// The direct_pinvoke HashTable keys are module names, and its values are HashTables
+// corresponding to entrypoint names within the module to be direct pinvoked.
+// A NULL value in the direct_pinvoke HashTable is understood to mean that all entrypoints
+// from the library are direct. It will overrule previously added HashTable of entrypoints,
+// and it will prevent new HashTable of entrypoints from being added.
+//
+// Processing:
+// The specified direct pinvoke, dpi, is ignored if it is empty or considered a comment.
+// It is then understood to be in the format of MODULE or MODULE!ENTRYPOINT.
+// A direct pinvoke in the form of MODULE is understood as enabling direct pinvoke for all
+// entrypoints within the particular module, and will override any previously added set
+// of entrypoint names.
+// A direct pinvoke in the form of MODULE!ENTRYPOINT is understood as enabling direct pinvoke
+// for the specific entrypoint in the module. It will not be added if the entire module
+// should be direct pinvoked, but otherwise will be added to a set of entrypoint names for
+// the particular module.
+//
+// Arguments:
+//  * acfg - the MonoAotCompiler instance
+//  * dpi (direct pinvoke) - the string passed in specifying a direct pinvoke
+//
+// Return Value:
+//  gboolean corresponding to whether the specified direct pinvoke was successfully
+//  processed (regardless of it being added or skipped) or if an error occurred.
+//
+//  Note - There are no extensive format checks, and is intended to behave akin to
+//  ConfigurablePInvokePolicy AddDirectPInvoke in NativeAOT
+//
+
+static gboolean
+process_specified_direct_pinvokes (MonoAotCompile *acfg, char *dpi)
+{
+	if (is_direct_pinvoke_parsable (dpi)) {
+		char *module;
+		char *entrypoint;
+		switch (parsed_direct_pinvoke (acfg, dpi, &module, &entrypoint)) {
+		case DIRECT_PINVOKE_ERR:
```

Comment:
Not sure how much value this enum does, wouldn't a gboolean be more simpler? I think the only case where you end up in default could probably be merged with the is_direct_pinvoke_parsable (failure trimming the string).

Reply:
```
gboolean result = FALSE;
char * direct_pinvoke = g_strdup (dpi);
if (direct_pinvoke && g_strstrip (direct_pinvoke )) {
    if (is_direct_pinvoke_parsable (direct_pinvoke)) {
        char * module, entrypoint;
        if (parsed_direct_pinvoke (acfg, direct_pinvoke, &module, &entrypoint))
            result = add_direct_pinvoke (acfg, &module, &entrypoint);
    } else {
        result = TRUE;
    }
}
g_free (direct_pinvoke);
return result;
```

You could also change input parameter to const char *dpi.

Reply:
Right, I originally had a boolean until I figured I needed to consider differentiating between skipping a direct pinvoke because it was empty after trimming white space and erroring from `g_strsplit`, `g_strdup`. But as you mentioned, if we relocate the trimming step, the differentiation no longer needs to happen.